sdc/datatypes/hpat_pandas_dataframe_functions.py

```diff
@@ -31,65 +31,45 @@
-    Pandas DataFrame method :meth:`pandas.DataFrame.count` implementation.
-
-    .. only:: developer
-
-        Test: python -m sdc.runtests sdc.tests.test_dataframe.TestDataFrame.test_count
-
-    Parameters
-    -----------
-    self: :class:`pandas.DataFrame`
-        input arg
-    axis:
-        *unsupported*
-    level:
-        *unsupported*
-    numeric_only:
-        *unsupported*
-
-    Returns
-    -------
-    :obj:`pandas.Series` or `pandas.DataFrame`
-            returns: For each column/row the number of non-NA/null entries. If level is specified returns a DataFrame.
-    """
-
-    _func_name = 'Method pandas.dataframe.count().'
-
-    if not isinstance(self, DataFrameType):
-        raise TypingError('{} The object must be a pandas.dataframe. Given: {}'.format(_func_name, self))
+from sdc.datatypes.hpat_pandas_series_functions import TypeChecker
 
-    if not (isinstance(axis, types.Omitted) or axis == 0):
-        raise TypingError("{} 'axis' unsupported. Given: {}".format(_func_name, axis))
+def sdc_pandas_dataframe_reduce_columns(df, name, params):
 
-    if not (isinstance(level, types.Omitted) or level is None):
-        raise TypingError("{} 'level' unsupported. Given: {}".format(_func_name, axis))
+    saved_columns = df.columns
+    data_args = tuple('data{}'.format(i) for i in range(len(saved_columns)))
+    all_params = ['df']
 
-    if not (isinstance(numeric_only, types.Omitted) or numeric_only is False):
-        raise TypingError("{} 'numeric_only' unsupported. Given: {}".format(_func_name, axis))
+    for key, value in params:
+        all_params.append('{}={}'.format(key, value))
+    ap = all_params.copy()
+    par = '{}'.format(', '.join(ap[1:]))
+    func_definition = 'def _reduce_impl({}):'.format(', '.join(all_params))
+    func_lines = [func_definition]
+    for i, d in enumerate(data_args):
+        line = '  {} = sdc.hiframes.api.init_series(sdc.hiframes.pd_dataframe_ext.get_dataframe_data(all_params[0], {}))'
+        func_lines.append(line.format(d + '_S', i))
+        func_lines.append(' {}_O = {}_S.{}({})'.format(d, d, name, par))
+    func_lines.append('  data = np.array(({},))'.format(
+        ", ".join(d + '_O' for d in data_args)))
+    func_lines.append('  index = sdc.str_arr_ext.StringArray(({},))'.format(
+        ', '.join('"{}"'.format(c) for c in saved_columns)))
+    func_lines.append('  return sdc.hiframes.api.init_series(data, index)')
+    loc_vars = {}
+    func_text = '\n'.join(func_lines)
 
-    def sdc_pandas_dataframe_count_impl(self, axis=0, level=None, numeric_only=False):
-        result_data = []
-        result_index = []
+    exec(func_text, {'sdc': sdc, 'np': numpy}, loc_vars)
+    _reduce_impl = loc_vars['_reduce_impl']
 
-        for dataframe_item in self._data:
-            item_count = dataframe_item.count()
-            item_name = dataframe_item._name
-            result_data.append(item_count)
-            result_index.append(item_name)
+    return _reduce_impl
```

Comment:
A small refactoring. But now it looks a little bit clearly:
```python
def _dataframe_reduce_columns_codegen(func_name, func_params, series_params, columns):
    result_name_list = []
    func_lines = [f'def {func_name}({func_params}):']
    for c in columns:
        result_d = f'result_{d}'
        func_lines += [f'  series_{d} = init_series(get_dataframe_data({all_params[0]}, {i}))',
                       f'  {result_d} = series_{d}.{func_name}({series_params})'
        result_name_list += result_d

    all_results = ', '.join(result_name_list)
    all_columns = ', '.join([f"'{c}'" for c in columns])

    func_lines += [f'  return pandas.Series([{all_results}], [{all_columns}])']
    func_text = '\n'.join(func_lines)

    global_vars = {'pandas': pandas, 'np': numpy, 
                   'init_series' : init_series, 
                   'get_dataframe_data': get_dataframe_data}

    return func_text, global_vars


def sdc_pandas_dataframe_reduce_columns(df, func_name, params):
    all_params = ['df']

    for key, value in params:
        all_params.append('{}={}'.format(key, value))
    ap = all_params.copy()
    par = '{}'.format(', '.join(ap[1:]))

    df_func_name = f'_df_{func_name}_impl'

    func_text, global_vars = _dataframe_reduce_columns_codegen(df_func_name, all_params, par, df.columns)
    loc_vars = {}

    exec(func_text, global_vars, loc_vars)
    _reduce_impl = loc_vars[df_func_name]

    return _reduce_impl
```

Please note. I haven't test it, so it is most likely full of bugs

I've split codegen part and eval part, so we can tests them separately and also each part become smaller and easier to understand.